feat(router): Add test helper for trigger navigations in tests

atscott opened the pull request:
In order to test components and services which depend on router navigations, such as `ActivatedRoute` instances, tests currently need to provide a fair bit of boilerplate _or_ they can set up a stub for `ActivatedRoute` and list it in the `providers` to override it in `TestBed`. This approach of stubbing the `ActivatedRoute` creates a situation that can easily cause the test to break. The stub often only mocks out the dependencies that the component/service _currently_ needs. This dependencies might change over time and break the test in an unexpected way. In addition, it is difficult to get the structure of `ActivatedRoute` exactly correct.

This change will allow unit tests to quickly set up routes, trigger real navigations in the Router, and get instances of component's to test along with real instances of `ActivatedRoute`. This all comes without needing to know that the component depends on `ActivatedRoute` at all. This becomes more important when considering that a component may be refactored in the future to use `@Input` rather than access data on the `ActivatedRoute` instance (see #18967). Tests which mock out `ActivatedRoute` would all break, but those which use `navigateForTest` would continue to work without needing any updates.

resolves #15779
resolves #48608

atscott committed:
feat(router): Add test helper for trigger navigations in tests

In order to test components and services which depend on router navigations, such as `ActivatedRoute` instances, tests currently need to provide a fair bit of boilerplate _or_ they can set up a stub for `ActivatedRoute` and list it in the `providers` to override it in `TestBed`. This approach of stubbing the `ActivatedRoute` creates a situation that can easily cause the test to break. The stub often only mocks out the dependencies that the component/service _currently_ needs. This dependencies might change over time and break the test in an unexpected way. In addition, it is difficult to get the structure of `ActivatedRoute` exactly correct.

This change will allow unit tests to quickly set up routes, trigger real navigations in the Router, and get instances of component's to test along with real instances of `ActivatedRoute`. This all comes without needing to know that the component depends on `ActivatedRoute` at all. This becomes more important when considering that a component may be refactored in the future to use `@Input` rather than access data on the `ActivatedRoute` instance (see #18967). Tests which mock out `ActivatedRoute` would all break, but those which use `navigateForTest` would continue to work without needing any updates.

resolves #15779
resolves #48608

AndrewKushnir commented on packages/router/testing/src/navigate_for_test.ts:
```diff
@@ -0,0 +1,84 @@
+ *     activated component of the `RouterOutlet`. If the outlet is not activated or a different
+ *     component is activated, this function will throw an error.
+ */
+export async function navigateForTest<T>(url: string, requiredRoutedComponentType: Type<T>):
```
I'm wondering if we should mimic the `navigateByUrl` function signature here and call it `navigateByUrlForTest` (bikeshedding :))? In this case the `instanceof` check would need to happen outside of the function:
```
const {activatedComponent} = await navigateByUrlForTest('a/b/c', { ... });
expect(activatedComponent).toBeInstanceOf(MyComponent);
```

Otherwise, the call would need a try-catch wrapper to translate an error to a Jasmine assertion or a `fail()` call (as demonstrated in the newly added tests below).

atscott replied:
The tests _do not_ need a `try...catch` but should instead use `expectAsync`. At the time I wrote the tests originally, I was too wiped to figure out why I couldn't get that working. They've now been fixed.

Resolving this conversation even though there are two threads. Maybe move the bikeshedding to a new comment? I don't feel like it's necessary to mirror `navigateByUrl` personally, but am open to bikeshedding nevertheless.

AndrewKushnir commented on packages/router/testing/src/navigate_for_test.ts:
```diff
@@ -0,0 +1,84 @@
+export async function navigateForTest<T>(url: string, requiredRoutedComponentType: Type<T>):
+    Promise<{hostFixture: ComponentFixture<{}>, activatedComponent: T}>;
+export async function navigateForTest<T>(url: string, requiredRoutedComponentType?: Type<T>):
+    Promise<{hostFixture: ComponentFixture<{}>, activatedComponent: T | null}> {
```
I think it'd be a bit tricky to explain what `hostFixture` is, since it represents a component that users don't have access to (and it was not provided by users). It looks more like an implementation detail which we expose. I'm wondering if it'd be enough to return a DebugElement that represents an activated component?

```
export async function navigateByUrlForTest(url: string, extras: NavigationBehaviorOptions): Promise<{activatedComponent: DebugElement|null}> {}
```

atscott replied:
We do need to return a fixture because it has things that may be required for the test, like `fixture.detectChanges`

AndrewKushnir commented on packages/router/testing/src/navigate_for_test.ts:
```diff
@@ -0,0 +1,84 @@
+    if (this.fixture !== undefined) {
+      return this.fixture;
+    }
+    if (!this.isCompilationDone) {
+      await TestBed.compileComponents();
+      this.isCompilationDone = true;
+    }
```
I think we may not need this, since typically the `TestBed.compileComponents()` call happens in a `beforeEach` block (as generated by the CLI):

```
  beforeEach(async () => {
    await TestBed.configureTestingModule({ ... }).compileComponents();
  });
```

So this is a part of the "setup" phase. I see the `navigateForTest` function more like the `TestBed.createComponent` function (but for a specific testing scenario), which relies on the setup phase to be completed.

atscott replied:
Removed. As long as the error message is clear when `compileComponents` is not already called, this should be alright.

AndrewKushnir reviewed (commented):
Looks great, just added a few comments.

atscott commented:
[presubmit](https://fusion2.corp.google.com/presubmit/tap/498200279/OCL:498200279:BASE:499903003:1672952707416:17f8768f/targets) with [local patch to BUILD](http://cl/498200279)

caretaker note: when syncing to g3, it will require a manual update to the BUILD file as shown in the linked CL.

jessicajaniuk commented on goldens/public-api/router/testing/index.md:
```diff
@@ -17,9 +18,22 @@ import { Router } from '@angular/router';
 
+// @public
+export function navigateForTest(url: string): Promise<{
+    hostFixture: ComponentFixture<{}>;
```
What does `host` mean in this context? Is it the host element? Seeing later that you're getting the rootFixture, does it make more sense to call it `rootFixture`?  It's just a bit unclear to me which component fixture this is referring to based on the name.

atscott replied:
`rootFixture` is also just a name (and not exposed externally). `host` vs `root` doesn't matter much to me so whichever you prefer.

jessicajaniuk replied:
I guess to clarify my question...what is the item that this is actually the fixture of. The name is a big vague.

atscott replied:
Yea, so I don't disagree but I don't know best to resolve this. Naming alone won't do it and I don't think we should expose the underlying component itself in the public API. I've updated the documentation on the method to include more explanation. Let me know if that's enough / what more you'd like to see.

jessicajaniuk replied:
Yeah, that documentation helps clarify it. I think that works perfectly. Thank you.

jessicajaniuk reviewed (approved):
reviewed-for: fw-testing, public-api

atscott committed:
refactor(router): Move `afterNextNavigation` helper function to a reusable location

There may be use-cases in the future where `afterNextNavigation` is used
in other locations. For example, a test helper function which navigates
should likely wait for all redirects to complete rather than simply
waiting for the promise returned from `navigate` to resolve.

devversion commented on packages/router/testing/test/BUILD.bazel:
```diff
@@ -0,0 +1,39 @@
+)
+
+# Tests rely on `async/await` for change detection.
+# This syntax needs to be downleveled until ZoneJS supports it.
```
Non-blocking note: please make sure this comment is actually true here. Until ZoneJS supports native async/await- this involves downleveling which is slower. So really the `zone_compatible_` variant should only be used if needed for change detection w/ ZoneJS and async/await

devversion reviewed (approved):
LGTM for bazel

atscott committed:
fixup! feat(router): Add test helper for trigger navigations in tests

jessicajaniuk reviewed (approved):
This is a great API change. I really like the `RouterTestingHarness` approach.

reviewed-for: fw-testing, public-api

AndrewKushnir commented on packages/router/testing/src/router_testing_harness.ts:
```diff
@@ -0,0 +1,151 @@
+   * The `RouterTestingHarness` also creates its own root component with a `RouterOutlet` for the
+   * purposes of rendering route components.
+   */
+  static async get(initialUrl?: string): Promise<RouterTestingHarness> {
```
Bikeshed: may be we can call it `create` instead, in which case a usage would look like this:
```
    const harness = await RouterTestingHarness.create();
    const activatedComponent = await harness.navigateByUrl('/', TestCmp);
```

atscott replied:
Because the same instance is reused within a test across multiple calls, I didn't want to call it just _create_. What about 
* `RouterTestingHarness.getOrCreate()`
* `RouterTestingHarness.obtain()`

AndrewKushnir replied:
Do we need to reuse an instance across tests? I think that was helpful with the previous API design (with a single function), but may be we can just include component creation into the `RouterTestingHarness.create` call instead and have a unique/clean version for each test?

atscott replied:
We want to reuse the same instance within the same `it`, not across different tests. Getting a new harness with a new component instance within the same test would be quite confusing.

AndrewKushnir replied:
Can we reuse the same instance like this? I.e. if we want to test multiple routes in a single test, we can create a harness once and perform multiple `navigateByUrl` calls.
```
    const harness = await RouterTestingHarness.create();
    const activatedComponent = await harness.navigateByUrl('/', TestCmp);

    // later in a test...
    const anotherActivatedComponent = await harness.navigateByUrl('/another-cmp', AnotherTestCmp);
```
Please let me know if I'm missing something.

atscott replied:
Yes, you _can_ and probably _would_ but what happens if you call `create` twice in the same test?

AndrewKushnir reviewed (commented):
@atscott looks great 👍 Just one minor proposal from the public API perspective.

atscott committed:
fixup! feat(router): Add test helper for trigger navigations in tests

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

jessicajaniuk commented:
This PR was merged into the repository by commit dedac8d3f73ebf4f05b773454e2a22ab5fa4bf7c.
